### ISSUE TITLE
Graph contrastive learning with augmentations

### ISSUE THREAD

**Issue by 2nazero** [AIML-K/GNN_Survey#20]

**Comment by 2nazero** [comment]
# Overall Summary
Contrastive learning with carefully designed data augmentations can significantly enhance the generalizability and robustness of Graph Neural Networks (GNNs) across various learning settings. GraphCL generates robust and transferable representations by applying four augmentation techniques—node dropping, edge perturbation, attribute masking, and subgraph sampling—each designed to incorporate specific prior knowledge about graph structure.

**Comment by 2nazero** [comment]
# Methodology
## Graph Data Augmentation
- **Node Dropping**: Randomly removes nodes, simulating situations where some vertices may be missing without altering the graph’s overall semantics.
- **Edge Perturbation**: Adds or removes edges, introducing robustness to connectivity changes.
- **Attribute Masking**: Masks certain node attributes, enforcing robustness against partial attribute loss.
- **Subgraph Extraction**: Samples a subgraph from $G$ using random walks, assuming that local structures can capture core semantic information.

## Graph Contrastive Learning Framework (GraphCL)
- **Graph Data Augmentation**: Create two correlated augmented views $\hat{G}_i$ and $\hat{G}_j$ for the same graph $G$, forming a positive pair.
- **GNN-based Encoder**: A GNN encoder $f(\cdot)$ extracts feature representations $h_i$ and $h_j$ for the augmented graphs $\hat{G}_i$ and $\hat{G}_j$.
- **Projection Head**: Maps these representations to another latent space using a non-linear transformation $g(\cdot)$, where the contrastive loss is computed. Typically implemented as a two-layer MLP.
- **Contrastive Loss Function**: Enforces similarity between positive pairs $(z_i, z_j)$ using a normalized temperature-scaled cross-entropy loss (NT-Xent) while distinguishing them from negative pairs (representations of different graphs). 

$$\ell_n = -\log \frac{\exp(\text{sim}(z_{n,i}, z_{n,j}) / \tau)}{\sum_{n' = 1}^{2N} \mathbb{1}_{[n' \neq n]} \exp(\text{sim}(z_{n,i}, z_{n',j}) / \tau)}$$

<img width="533" alt="스크린샷 2024-11-06 오후 4 35 37" src="https://github.com/user-attachments/assets/0296ebaa-b903-4a83-99fd-8b71af30e78d">

**Comment by 2nazero** [comment]
# Experiment
## Semi-supervised Learning
- **Objective**: Evaluate GraphCL in a semi-supervised setting, particularly on graph classification tasks.
- **Setup**: Conducted on the TUDataset benchmark, comparing GraphCL with other pre-training methods, such as **GAE** (Graph Autoencoder for adjacency reconstruction) and **Infomax** (local & global representation consistency).
- **Results**: GraphCL outperforms traditional pre-training methods and shows improved accuracy over both training from scratch and using augmentation alone.

## 2. Unsupervised Representation Learning
- **Objective**: Test the quality of GraphCL’s learned representations in an unsupervised setup, using them for downstream classification tasks.
- **Setup**: Graph representations learned by GraphCL are fed into an SVM classifier, with comparisons against state-of-the-art methods, including **graph kernels** (e.g., Graphlet Kernel, Weisfeiler-Lehman kernel) and other unsupervised learning approaches like **node2vec** and **InfoGraph**.
- **Results**: GraphCL achieves state-of-the-art performance on most datasets, particularly in large-scale graphs, demonstrating robust unsupervised learning capabilities.

**Comment by 2nazero** [comment]
# Results
## Semi-supervised Learning
- **Objective**: Evaluate GraphCL in a semi-supervised setting, particularly on graph classification tasks.
- **Setup**: Conducted on the TUDataset benchmark, comparing GraphCL with other pre-training methods, such as **GAE** (Graph Autoencoder for adjacency reconstruction) and **Infomax** (local & global representation consistency).
- **Results**: GraphCL outperforms traditional pre-training methods and shows improved accuracy over both training from scratch and using augmentation alone.
<img width="526" alt="스크린샷 2024-11-06 오후 4 47 27" src="https://github.com/user-attachments/assets/dd0e2516-6982-47a6-b069-8e0eb311466b">


## Unsupervised Representation Learning
- **Objective**: Test the quality of GraphCL’s learned representations in an unsupervised setup, using them for downstream classification tasks.
- **Setup**: Graph representations learned by GraphCL are fed into an SVM classifier, with comparisons against state-of-the-art methods, including **graph kernels** (e.g., Graphlet Kernel, Weisfeiler-Lehman kernel) and other unsupervised learning approaches like **node2vec** and **InfoGraph**.
- **Results**: GraphCL achieves state-of-the-art performance on most datasets, particularly in large-scale graphs, demonstrating robust unsupervised learning capabilities.
<img width="527" alt="스크린샷 2024-11-06 오후 4 47 42" src="https://github.com/user-attachments/assets/cea551a4-f59c-43d8-ad27-1e7a38bbd616">



## Transfer Learning
- **Objective**: Evaluate the transferability of GraphCL by pre-training on one dataset and fine-tuning on another, simulating real-world applications.
- **Setup**: Transfer learning is tested on molecular and protein datasets, with comparisons to other pre-training schemes such as **EdgePred** and **ContextPred**.
- **Results**: GraphCL shows top performance on 5 out of 9 datasets, demonstrating effectiveness in transferring learned representations across different domains.
<img width="531" alt="스크린샷 2024-11-06 오후 4 47 55" src="https://github.com/user-attachments/assets/215b58aa-b1d2-4310-8ebd-d6ae441e238f">




## Adversarial Robustness
- **Objective**: Assess GraphCL's robustness under adversarial attacks, which aim to mislead the GNN model by manipulating graph structures or node features.
- **Setup**: Experiments are conducted with synthetic datasets under three attack types—**RandSampling**, **GradArgmax**, and **RL-S2V**—using the Structure2vec encoder.
- **Results**: GraphCL significantly improves robustness, especially in deeper GNN models, demonstrating resistance to various attack strategies.
<img width="518" alt="스크린샷 2024-11-06 오후 4 48 12" src="https://github.com/user-attachments/assets/aabba71a-4567-44d4-80c0-cbeaa3440cb7">

**Comment by 2nazero** [comment]
# Code Comparison 1

```
root@7f02d7df1074:~/pretrain-gnns/bio# sh finetune_tune.sh 0 0
BioDataset(88000)
species splitting
Data(center_node_idx=[1], edge_attr=[1328, 9], edge_index=[2, 1328], go_target_downstream=[40], go_target_pretrain=[5000], species_id=[1], x=[96, 1])
====epoch 1
Iteration: 100%|█████████████████████████████████████████████████████████████| 1898/1898 [30:50<00:00,  1.37it/s]
====Evaluation
ommitting training evaluation
Iteration: 100%|█████████████████████████████████████████████████████████████████| 34/34 [03:02<00:00,  4.33s/it]
Iteration: 100%|█████████████████████████████████████████████████████████████████| 26/26 [03:06<00:00,  6.88s/it]
Iteration: 100%|█████████████████████████████████████████████████████████████████| 26/26 [03:02<00:00,  7.07s/it]
[0.64972875 0.56910686 0.62580789 0.6868586  0.70972239 0.61607357
 0.96744478 0.65714751 0.76729728 0.64269275 0.36718232 0.59988434
 0.76854851 0.79896305 0.69637291 0.73559006 0.6387814  0.63413933
 0.78645537 0.58860739 0.82837439 0.71741512 0.6722444  0.76553147
 0.77369004 0.79815178 0.82105677 0.71426084 0.7525173  0.81041924
 0.75561212 0.7137282  0.68113799 0.55293587 0.85315147 0.51297131
 0.65642465 0.6613931  0.85084065 0.66751868]
[0.65700924 0.5555769  0.62989931 0.71378914 0.71114342 0.61818688
 0.98854888 0.62500914 0.74383113 0.64056977 0.38747991 0.63407911
 0.78000789 0.77722679 0.6555413  0.83901423 0.70693404 0.69227741
 0.78826231 0.68094605 0.84529519 0.90383051 0.59848045 0.74160788
 0.71667653 0.70890729 0.95850472 0.63821104 0.74069293 0.83942561
 0.75260394 0.68786503 0.74654336 0.71170599 0.85618678 0.52467922
 0.73946938 0.58384304 0.61967805 0.62245365]
...
```
